modules/stochastic_tools/src/samplers/LatinHypercubeSampler.C

```diff
@@ -0,0 +1,109 @@
+  params.addClassDescription("Latin Hypercube Sampler.");
+  params.addRequiredParam<dof_id_type>("num_rows",
+                                       "The number of rows per matrix to generate; a single value "
+                                       "or one value per distribution may be supplied.");
+  params.addRequiredParam<std::vector<DistributionName>>(
+      "distributions",
+      "The distribution names to be sampled, the number of distributions provided defines the "
+      "number of columns per matrix.");
+  params.addRequiredParam<std::vector<unsigned int>>(
+      "num_bins", "The number of intervals to consider within the sampling.");
+  params.addRequiredParam<std::vector<double>>(
+      "upper_limits", "The lower limit of probability for each of the associated distributions.");
+  params.addRequiredParam<std::vector<double>>(
+      "lower_limits", "The upper limit of probability for each of the associated distributions.");
+  return params;
+}
+
+LatinHypercubeSampler::LatinHypercubeSampler(const InputParameters & parameters)
+  : Sampler(parameters),
+    _distribution_names(getParam<std::vector<DistributionName>>("distributions")),
+    _num_bins_input(getParam<std::vector<unsigned int>>("num_bins")),
+    _upper_limits(getParam<std::vector<double>>("upper_limits")),
+    _lower_limits(getParam<std::vector<double>>("lower_limits"))
+{
+  for (const DistributionName & name : _distribution_names)
+    _distributions.push_back(&getDistributionByName(name));
+
+  if (_upper_limits.size() != _distributions.size())
+    paramError("upper_limits",
+               "The length of 'upper_limits' must match the length of 'distributions'");
+
+  if (_lower_limits.size() != _distributions.size())
+    paramError("lower_limits",
+               "The length of 'lower_limits' must match the length of 'distributions'");
+
+  if (_num_bins_input.size() != 1 && _num_bins_input.size() != _distributions.size())
+    paramError("num_bins",
+               "The length of 'num_bins' must be one or match the length of 'distributions'");
+
+  for (const auto & lim : _upper_limits)
+    if (lim < 0 || lim > 1)
+      paramError("upper_limits", "The items in 'upper_limits' must be in the range [0,1]");
+
+  for (const auto & lim : _lower_limits)
+    if (lim < 0 || lim > 1)
+      paramError("lower_limits", "The items in 'lower_limits' must be in the range [0,1]");
+
+  setNumberOfRandomSeeds(2); // 0 = Bin seed; 1 = distribution sample seed
+  setNumberOfRows(getParam<dof_id_type>("num_rows"));
+  setNumberOfCols(_distributions.size());
+}
+
+void
+LatinHypercubeSampler::sampleSetUp()
+{
+  // Setup bin information
+  const std::size_t num_dists = _distributions.size();
+  _size_bins.resize(num_dists);
+  _num_bins = _num_bins_input.size() == 1 ? std::vector<unsigned int>(_num_bins_input[0], num_dists)
+                                          : _num_bins_input;
+
+  for (std::size_t dist_idx = 0; dist_idx < num_dists; ++dist_idx)
```

Comment:
use range-based for loop here?

Reply:
No, the index is used to access two different lists.